src/mountinfo/mountinfo.c

```diff
@@ -356,6 +381,157 @@ find_mounts(struct args *args)
+}
+
+static int is_mounted(const char *mntpath)
+{
+	size_t num_mounts = 0;
+	struct args args = { .process = check_is_mounted, .check_mntpath = mntpath };
+	RC_STRINGLIST *l = find_mounts(&args, &num_mounts);
+	rc_stringlist_free(l);
+	return num_mounts > 0;
+}
+
+static pid_t run_umount(const char *mntpath,
+	const char **umount_args, int umount_args_num)
+{
+	struct exec_args args;
+	struct exec_result res;
+	const char *argv[UMOUNT_ARGS_MAX + 3];
+	int k, i = 0;
+
+	argv[i++] = "umount";
+	for (k = 0; k < umount_args_num; ++k)
+		argv[i++] = umount_args[k];
+	argv[i++] = mntpath;
+	argv[i++] = NULL;
+
+	args = exec_init(argv);
+	args.redirect_stdout = args.redirect_stderr = EXEC_DEVNULL;
+	res = do_exec(&args);
+	if (res.pid < 0)
+		eerrorx("%s: failed to run umount: %s", applet, strerror(errno));
+	return res.pid;
+}
+
+static void fuser_run(struct run_queue *rp, const char *fuser_opt)
+{
+	static int fuser_exec_failed = 0;
+	const char *argv[] = { "fuser", fuser_opt, rp->mntpath, NULL };
+	struct exec_result res;
+	struct exec_args args;
+
+	/* if exec failed, fuser likely doesn't exist. so don't retry */
+	if (fuser_exec_failed)
+		return;
+
+	args = exec_init(argv);
+	args.redirect_stdout = EXEC_MKPIPE;
+	args.redirect_stderr = EXEC_DEVNULL;
+	res = do_exec(&args);
+	if (res.pid < 0) {
+		fuser_exec_failed = 1;
+	} else {
+		rp->fuser_pid = res.pid;
+		rp->fuser_stdoutfd = res.proc_stdout;
+		rp->fuser_exec_time = tm_now();
+	}
+}
+
+static int fuser_decide(struct run_queue *rp,
+	const char *fuser_opt, const char *fuser_kill_prefix)
+{
+	char buf[1<<12];
+	char selfpid[64];
+	int read_maybe_truncated;
+	ssize_t n;
+
+	if (rp->fuser_stdoutfd < 0)
+		return 0;
+
+	buf[0] = ' ';
+	n = read(rp->fuser_stdoutfd, buf + 1, sizeof buf - 3);
+	close(rp->fuser_stdoutfd);
+	rp->fuser_stdoutfd = -1;
+	read_maybe_truncated = (n == sizeof buf - 3);
+	if (n < 0 || read_maybe_truncated)
+		return 0;
+	while (n > 0 && buf[n] == '\n')
+		--n;
+	buf[n+1] = ' ';
+	buf[n+2] = '\0';
+	snprintf(selfpid, sizeof selfpid, " %lld ", (long long)getpid());
+
+	if (strstr(buf, selfpid)) {
+		/* lets not kill ourselves */
+		eerror("Unmounting %s failed because we are using it", rp->mntpath);
+		return -1;
+	} else if (strcmp(buf, "  ") == 0) {
+		if (rp->try_count >= TRY_MAX) {
+			eerror("Unmounting %s failed but fuser finds no one using it", rp->mntpath);
+			return -1;
+		}
+		/* it's possible that whatever was using the mount stopped
+		 * using it now, so allow 1 more retry */
+		rp->try_count = TRY_MAX;
+		return 0;
+	} else {
+		char sig[32];
+		const char *argv[] = {
+			"fuser", sig, "-k", fuser_opt, rp->mntpath, NULL
+		};
+		struct exec_result res;
+		struct exec_args args = exec_init(argv);
+		args.redirect_stdout = args.redirect_stderr = EXEC_DEVNULL;
+		snprintf(sig, sizeof sig, "%s%s", fuser_kill_prefix,
+			rp->try_count == TRY_MAX ? "KILL" : "TERM");
+		res = do_exec(&args);
+		if (res.pid > 0)
+			rc_waitpid(res.pid);
```

Comment:
The first fuser command to query the pids has a timeout and is async, this one to send the signals on the other hand is synchronous and has no timeout (copied from the `rc-mount.sh` script).

But this doesn't seem to make sense (?) If we anticipate fuser might take a while then shouldn't we be timing out *both* of the calls rather than just one of them?

Also the only reason we're doing 2 commands (query pids + send signal) is to avoid killing ourselves if we happen to be using that mount. If we can figure out whether we're using it or not beforehand then it can be changed into a single fuser call which directly sends the signal (async, with timeout).